aws-s3-transfer-manager/src/io/adapters.rs

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::io::stream::StreamContext;
+use crate::io::Buffer;
+use crate::io::PartStream;
+use crate::io::SizeHint;
+
+use crate::io::PartData;
+
+use std::mem;
+use std::task::Poll;
+
+use bytes::Bytes;
+use pin_project_lite::pin_project;
+
+// TODO - add feature gate for this?
+pin_project! {
+    /// A wrapper that implements [`PartStream`] trait for an inner type
+    /// that implements Tokio's IO traits.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use aws_s3_transfer_manager::io::adapters::TokioIo;
+    /// use aws_s3_transfer_manager::io::{InputStream, SizeHint};
+    /// use tokio::io::AsyncRead;
+    ///
+    /// fn into_input_stream<T: AsyncRead + Send + Sync + 'static>(inner: T, content_length: u64) -> InputStream {
+    ///     InputStream::from_part_stream(TokioIo::new(inner, SizeHint::exact(content_length)))
+    /// }
+    /// ```
+    #[derive(Debug)]
+    pub struct TokioIo<T> {
```

Comment:
No. It's generally more flexible to leave the bounds on the trait implementations. This allows for a type to potentially be more broadly applicable as it can implement multiple traits for some type `T` depending on what `T` is.

Reply:
Relevant: https://stackoverflow.com/questions/49229332/should-trait-bounds-be-duplicated-in-struct-and-impl